spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -603,101 +636,239 @@ private void invokeBatchListener(final ConsumerRecords<K, V> records) {
-					}
-					catch (Error er) { //NOSONAR
-						this.logger.error("Error handler threw an error", er);
-						throw er;
+				catch (RuntimeException e) {
+					this.logger.error("Transaction rolled back", e);
+					this.consumer.seek(new TopicPartition(record.topic(), record.partition()), record.offset());
+					break;
+				}
+			}
+		}
+
+		private void doInvokeWithRecords(final ConsumerRecords<K, V> records) throws Error {
+			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+			while (iterator.hasNext()) {
+				final ConsumerRecord<K, V> record = iterator.next();
+				if (this.logger.isTraceEnabled()) {
+					this.logger.trace("Processing " + record);
+				}
+				doInvokeRecordListener(record, null);
+			}
+		}
+
+		/**
+		 * Actually invoke the listner.
+		 * @param record the record.
+		 * @param producer the producer - only if we're running in a transaction, null
+		 * otherwise.
+		 * @return an exception.
+		 * @throws Error an error.
+		 */
+		private RuntimeException doInvokeRecordListener(final ConsumerRecord<K, V> record,
+				@SuppressWarnings("rawtypes") Producer producer) throws Error {
+			try {
+				switch (this.listenerType) {
+					case ACKNOWLEDGING_CONSUMER_AWARE:
+						this.listener.onMessage(record,
+								this.isAnyManualAck
+										? new ConsumerAcknowledgment(record)
+										: null, this.consumer);
+						break;
+					case CONSUMER_AWARE:
+						this.listener.onMessage(record, this.consumer);
+						break;
+					case ACKNOWLEDGING:
+						this.listener.onMessage(record,
+								this.isAnyManualAck
+										? new ConsumerAcknowledgment(record)
+										: null);
+						break;
+					case SIMPLE:
+						this.listener.onMessage(record);
+						break;
+				}
+				if (!this.isAnyManualAck && !this.autoCommit) {
+					this.acks.add(record);
+				}
+				if (producer != null) {
+					sendOffsetsToTransaction(producer);
+				}
+			}
+			catch (RuntimeException e) {
+				if (this.containerProperties.isAckOnError() && !this.autoCommit && producer == null) {
+					this.acks.add(record);
+				}
+				if (this.errorHandler == null) {
+					throw e;
+				}
+				try {
+					this.errorHandler.handle(e, record, this.consumer);
+					if (producer != null) {
+						try {
+							sendOffsetsToTransaction(producer);
+						}
+						catch (Exception e1) {
+							this.logger.error("Send offsets to transaction failed", e1);
+						}
 					}
 				}
+				catch (RuntimeException ee) {
+					this.logger.error("Error handler threw an exception", ee);
+					return ee;
+				}
+				catch (Error er) { //NOSONAR
+					this.logger.error("Error handler threw an error", er);
+					throw er;
+				}
 			}
+			return null;
+		}
+
+		@SuppressWarnings({ "unchecked", "rawtypes" })
+		protected void sendOffsetsToTransaction(Producer producer) {
+			handleAcks();
+			Map<TopicPartition, OffsetAndMetadata> commits = buildCommits();
+			producer.sendOffsetsToTransaction(commits, this.consumerGroupId);
```

Comment:
Well, I'm talking here about error use-case.
We send `acks` for commit only if we have `ackOnError`, but we call this method in case of exception unconditionally. Therefore when `ackOnError == false`, the `commits` is empty.